src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinding.cs

```diff
@@ -1,32 +1,37 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
+using System.Collections.Frozen;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public sealed class TagHelperBinding
+internal sealed class TagHelperBinding
 {
+    public string TagName { get; }
+    public string? ParentTagName { get; }
+    public ImmutableArray<KeyValuePair<string, string>> Attributes { get; }
+    public FrozenDictionary<TagHelperDescriptor, ImmutableArray<TagMatchingRuleDescriptor>> Mappings { get; }
```

Comment:
Why FrozenDictionary and not ImmutableDictionary?

Reply:
Because `ImmutableDictionary<,>` lookups are _slow_. `ImmutableDictionary<,>` is intended useful for when you need to perform non-destructive mutation that reuses the contents of previous snapshots. Because of that, the internals of `ImmutableDictionary<,>` use a tree structure, which makes lookups essentially O(log N) rather than O(1). `FrozenDictionary` is designed for fast lookups and enumeration of data that doesn't need to change.